core/src/collections/DynamicValue.hpp

```diff
@@ -28,47 +28,130 @@
+            DynamicValue& operator=(const std::vector<uint8_t>& rhs);
+            /// Change the value to a std::vector<uint8_t> by moving in.
+            DynamicValue& operator=(std::vector<uint8_t>&& rhs);
+
+            /// Change the value to a bool by copying.
+            DynamicValue& operator=(bool rhs);
+
+            /// Change the value to a int32_t by copying.
+            DynamicValue& operator=(int32_t rhs);
+
+            /// Change the value to a int64_t by copying.
+            DynamicValue& operator=(int64_t rhs);
+
+            /// Change the value to a double by copying.
+            DynamicValue& operator=(double rhs);
+
+            /// Change the value to a DynamicArray by copying.
+            DynamicValue& operator=(const std::shared_ptr<DynamicArray>& rhs);
 
+            /// Change the value to a DynamicObject by copying.
+            DynamicValue& operator=(const std::shared_ptr<DynamicObject>& rhs);
+
+            /// Try to get the value as if it were a std::string.
+            optional<std::string&> asStr();
+            /// Try to get the value as if it were a std::string.
+            optional<const std::string&> asStr() const;
+
+            /// Try to get the value as if it were a std::vector<uint8_t>.
+            optional<std::vector<uint8_t>&> asData();
+            /// Try to get the value as if it were a std::vector<uint8_t>.
+            optional<const std::vector<uint8_t>&> asData() const;
+
+            /// Try to get the value as if it were a bool.
+            optional<bool&> asBool();
+            /// Try to get the value as if it were a bool.
+            optional<bool> asBool() const;
+
+            /// Try to get the value as if it were a int32_t.
+            optional<int32_t&> asInt32();
+            /// Try to get the value as if it were a int32_t.
+            optional<int32_t> asInt32() const;
+
+            /// Try to get the value as if it were a int64_t.
+            optional<int64_t&> asInt64();
+            /// Try to get the value as if it were a int64_t.
+            optional<int64_t> asInt64() const;
+
+            /// Try to get the value as if it were a double.
+            optional<double&> asDouble();
+            /// Try to get the value as if it were a double.
+            optional<double> asDouble() const;
+
+            /// Try to get the value as if it were a DynamicArray.
+            optional<std::shared_ptr<DynamicArray>&> asArray();
+            /// Try to get the value as if it were a DynamicArray.
+            optional<const std::shared_ptr<DynamicArray>&> asArray() const;
+
+            /// Try to get the value as if it were a DynamicObject.
+            optional<std::shared_ptr<DynamicObject>&> asObject();
+            /// Try to get the value as if it were a DynamicObject.
+            optional<const std::shared_ptr<DynamicObject>&> asObject() const;
+
+        private:
+            // Tag representing the current in-use variant
+            api::DynamicType type;
```

Comment:
Can remove this as well.
define values inside variant in the same order as in DynamicType, for UNDEFINED create a dummy struct Undefined{}; and then use boost::variant::which with conversion to enum.